### PR TITLE
feat: action server for initial pose MOVEMENT-1231

### PR DESCRIPTION
**Pull request by 27Apoorva**

### CODE REVIEW

File: amcl/src/amcl_node.cpp [hunk 1]
```diff
@@ -477,6 +497,37 @@ AmclNode::AmclNode() :
   laser_check_interval_ = ros::Duration(15.0);
   check_laser_timer_ = nh_.createTimer(laser_check_interval_,
                                        boost::bind(&AmclNode::checkLaserReceived, this, _1));
+
+  // publish ready signal
+  ready_pub_.publish(true);
```
Comment (mbittarelli6river) [inline amcl/src/amcl_node.cpp]: same comment as other repo about maybe either doing this as the last thing in this function or move it to another method entirely so this object has finished constructed by the time it's published

File: amcl/src/amcl_node.cpp [hunk 2]
```diff
@@ -477,6 +497,37 @@ AmclNode::AmclNode() :
...
+}
+
+void AmclNode::executeInitialPoseCB(const move_base_msgs::SetInitialPoseGoalConstPtr &goal)
+{
+  ROS_INFO("Executing, inital pose server action");
+  if(use_emulator_)
+  {
+    driver_pose_pub_.publish(goal->initialPose);
+    ROS_INFO("Sleeping for brainstem to publish new tf/odometry");
+    ros::Duration(1.0).sleep();
```
Comment (mbittarelli6river) [inline amcl/src/amcl_node.cpp:525]: magic sleep numbers? this seems like possible dangerous async behavior. what happens if we didnt sleep? maybe this should be a polling action with a timeout instead?

File: amcl/src/amcl_node.cpp [hunk 3]
```diff
@@ -249,12 +259,15 @@ class AmclNode
...
     ros::Publisher particlecloud_pub_;
     ros::Publisher invalid_pose_percent_pub_;
+    ros::Publisher ready_pub_;
     ros::ServiceServer global_loc_srv_;
     ros::ServiceServer nomotion_update_srv_; //to let amcl update samples without requiring motion
     ros::ServiceServer set_map_srv_;
     ros::Subscriber initial_pose_sub_old_;
     ros::Subscriber map_sub_;
 
+    ros::Publisher driver_pose_pub_;
```
Comment (dnikitop) [inline amcl/src/amcl_node.cpp]: can we make it brainstem_driver_pose_pub

File: amcl/src/amcl_node.cpp (hunk 2 above)
Comment (27Apoorva) [reply amcl/src/amcl_node.cpp:525]: yeah so first thing is that this sleep is needed only in sim/emulator mode. Basically amcl needs to have updated odometery from brainstem. and if we don't sleep the amcl position doesn't get updated. there are other places in code where we do this as well. probably should update all of that as well

File: amcl/src/amcl_node.cpp [hunk 4]
```diff
@@ -477,6 +499,43 @@ AmclNode::AmclNode() :
...
+  // publish ready signal
+  ready_pub_.publish(signal);
+}
+
+void AmclNode::executeInitialPoseCB(const move_base_msgs::SetInitialPoseGoalConstPtr &goal)
+{
+  ROS_INFO("Executing, inital pose server action");
+  if(use_emulator_)
+  {
+    brainstem_driver_pose_pub_.publish(goal->initialPose);
```
Comment (dnikitop) [inline amcl/src/amcl_node.cpp:523]: so if emulation mode is enabled, the brainstem driver looks at a different topic. Is there some diagram to show what this all is doing and how it changed.

Comment (27Apoorva) [reply amcl/src/amcl_node.cpp:523]: if emulation is on the odomtery doesn't come from brainstem emulator so we don't need to publish on this topic to republish new tfs. I can document it in wiki page for this feature

File: amcl/src/amcl_node.cpp (hunk 2 above)
Comment (ccpowers) [reply amcl/src/amcl_node.cpp:525]: does this work with a shorter duration? seems like it maybe just needs the publisher to spin once? it's ok for now

Comment (27Apoorva) [reply amcl/src/amcl_node.cpp:525]: this is how much we wait when we use `2d position` in RVIz

**Review by mbittarelli6river — APPROVED** [review]

**Pull request merged**